test/hotspot/jtreg/serviceability/jvmti/RedefineClasses/RedefineRetransform/libRedefineRetransform.cpp

```diff
@@ -0,0 +1,277 @@
+                jvmti->Deallocate(buf);
+                return;
+            }
+
+            memcpy(buf, arrayPtr, len);
+
+            jni_env->ReleaseByteArrayElements(classLoadHookNewClassBytes, arrayPtr, JNI_ABORT);
+
+            *new_class_data = buf;
+            *new_class_data_len = len;
+
+            _log("  ClassFileLoadHook: set new class bytes\n");
+        }
+        _log("<<ClassFileLoadHook\n");
+    }
+}
+
+JNIEXPORT jint JNICALL Agent_OnLoad(JavaVM* jvm, char* options, void* reserved) {
+    jint res = jvm->GetEnv((void **)&jvmti, JVMTI_VERSION_1_1);
+    if (res != JNI_OK) {
+        _log("Failed to get JVMTI interface: %ld\n", res);
+        return JNI_ERR;
+    }
+
+    jvmtiCapabilities caps;
+    memset(&caps, 0, sizeof(caps));
+
+    caps.can_redefine_classes = 1;
+    caps.can_retransform_classes = 1;
+    jvmti->AddCapabilities(&caps);
+
+    jvmtiEventCallbacks eventCallbacks;
+    memset(&eventCallbacks, 0, sizeof(eventCallbacks));
+    eventCallbacks.ClassFileLoadHook = callbackClassFileLoadHook;
+    res = jvmti->SetEventCallbacks(&eventCallbacks, sizeof(eventCallbacks));
+    if (res != JVMTI_ERROR_NONE) {
+        _log("Error setting event callbacks: %ld\n", res);
+        return JNI_ERR;
+    }
+
+    return JNI_OK;
+}
+
+JNIEXPORT void JNICALL
+Agent_OnUnload(JavaVM* jvm) {
+    return;
+}
+
+
+JNIEXPORT jbyteArray JNICALL
+Java_RedefineRetransform_nRedefine(JNIEnv* env, jclass klass,
+                                   jclass testClass, jbyteArray classBytes, jbyteArray classLoadHookBytes) {
+
+    _log(">>nRedefine\n");
+    jsize len = env->GetArrayLength(classBytes);
+    jbyte* arrayPtr = env->GetByteArrayElements(classBytes, nullptr);
+    if (arrayPtr == nullptr) {
+        _log("nRedefine: Failed to get array elements\n");
+        return nullptr;
+    }
+    if (classLoadHookBytes != nullptr) {
+        classLoadHookNewClassBytes = (jbyteArray)env->NewGlobalRef(classLoadHookBytes);
+    }
+
+
+    jvmtiError err = jvmti->SetEventNotificationMode(JVMTI_ENABLE, JVMTI_EVENT_CLASS_FILE_LOAD_HOOK, nullptr);
+    if (err != JVMTI_ERROR_NONE) {
+        env->ReleaseByteArrayElements(classBytes, arrayPtr, JNI_ABORT);
+        _log("nRedefine: SetEventNotificationMode(JVMTI_ENABLE) error %d\n", err);
+        return nullptr;
+    }
+
+    jvmtiClassDefinition classDef;
+    memset(&classDef, 0, sizeof(classDef));
+    classDef.klass = testClass;
+    classDef.class_byte_count = len;
+    classDef.class_bytes = (unsigned char *)arrayPtr;
+
+    jvmtiError err2 = jvmti->RedefineClasses(1, &classDef);
+
+    if (err2 != JVMTI_ERROR_NONE) {
+        _log("nRedefine: RedefineClasses error %d", err2);
+        // don't exit here, need to cleanup
+    }
+
+    err = jvmti->SetEventNotificationMode(JVMTI_DISABLE, JVMTI_EVENT_CLASS_FILE_LOAD_HOOK, nullptr);
+    if (err != JVMTI_ERROR_NONE) {
+        _log("nRedefine: SetEventNotificationMode(JVMTI_DISABLE) error %d\n", err);
+    }
```

Comment:
It makes sense to introduce an utility function which does SetEventNotificationMode.
It can be two separate functions to enable and disable or one single function can support both cases.

Reply:
Removed code duplication - now the code is in helper class